servicenow/servicenow-mid-server.lua

```diff
@@ -0,0 +1,459 @@
+      os.execute("sleep " .. tonumber(1))
+      return queue:flush()
+    end
+
+    -- Here come the filters
+    -- Host Status/Service Status only
+    if not (e.category == 1 and (e.element == 24 or e.element == 14)) then
+      broker_log:info(3, "write: Neither host nor service status event. Dropping.")
+      return true
+    end
+
+    -- on drop les meta services pour le moment
+    if not e.host_id then
+        return true
+    end
+
+    if not e.host_id and not e.output:find("Meta-Service") == 1 then
+        broker_log:error(1, "write: Event has no host_id: " .. broker.json_encode(e))
+        return true
+    end
+
+    -- workaround https://github.com/centreon/centreon-broker/issues/201
+    current_event = broker.json_encode(e)
+    broker_log:info(3, "write: Raw event: " .. current_event)
+
+    -- Ignore objects in downtime
+    if e.scheduled_downtime_depth ~= 0 then --we keep only events in hard state and not in downtime -- Modif du 18/02/2020 => Simon Bomm
+      broker_log:info(3, "write: Scheduled downtime. Dropping.")
+      return true
+    end
+
+    -- Ignore objects with notifications disabled
+    if e.notify == false then
+      broker_log:info(3, "write: Notifications not enabled. Dropping.")
+      return true
+    end
+
+    -- Ignore SOFT 
+    if e.state_type and e.state_type ~= 1 then
+      broker_log:info(3, "write: " .. e.host_id .. "_" .. ifnil_or_empty(e.service_id, "H") .. " Not HARD state type. Dropping.")
+      return true
+    end
+
+    -- Ignore states different from previous hard state only
+    if e.last_hard_state_change and e.last_check and e.last_hard_state_change < e.last_check then
+      broker_log:info(3, "write: " .. e.host_id .. "_" .. ifnil_or_empty(e.service_id, "H") .. " Last hard state change prior to last check => no state change. Dropping.")
+      return true
+    end
+
+    -- workaround https://github.com/centreon/centreon-broker/issues/201
+    if current_event == previous_event then
+      broker_log:info(3, "write: Duplicate event ignored.")
+      return true
+    end
+
+    -- Ignore pending states
+    if e.state and e.state == 4 then
+      broker_log:info(3, "write: " .. e.host_id .. "_" .. ifnil_or_empty(e.service_id, "H") .. " Pending state ignored. Dropping.")
+      return true
+    end
+
+    -- The current event now becomes the previous
+    previous_event = current_event
+    -- Once all the filters have been passed successfully, we can add the current event to the queue
+    queue:add(e)
+
+    -- Then we check whether it is time to send the events to the receiver and flush
+    if (#queue.events >= queue.max_buffer_size) then
+      broker_log:info(2, "write: Queue max size (" .. #queue.events .. "/" .. queue.max_buffer_size .. ") is reached, flushing data")
+      return queue:flush()
+    end
+    broker_log:info(3, "write: Ending function")
+
+    return true
+end
+
```

Comment:
I recommend to add a function flush like this:
```
function flush()
  return queue:flush()
end
```

If not added, in case of retention files, the write() function won't be called, broker will only try to call the flush(). But if not implemented, nothing will happen and the retention will continue to grow up.